python/grass/script/setup.py

```diff
@@ -412,31 +414,50 @@ def finish(self):
         if not self.active:
             raise ValueError("Attempt to finish an already finished session")
         self._active = False
-        finish()
+        finish(start_time=self._start_time)
 
 
 # clean-up functions when terminating a GRASS session
 # these fns can only be called within a valid GRASS session
-def clean_default_db():
-    # clean the default db if it is sqlite
-    from grass.script import core as gcore
-    from grass.script import db as gdb
 
-    conn = gdb.db_connection()
-    if conn and conn["driver"] == "sqlite":
-        # check if db exists
-        gisenv = gcore.gisenv()
-        database = conn["database"]
-        database = database.replace("$GISDBASE", gisenv["GISDBASE"])
-        database = database.replace("$LOCATION_NAME", gisenv["LOCATION_NAME"])
-        database = database.replace("$MAPSET", gisenv["MAPSET"])
-        if os.path.exists(database):
-            gcore.message(_("Cleaning up default sqlite database ..."))
-            gcore.start_command("db.execute", sql="VACUUM")
-            # give it some time to start
-            import time
 
-            time.sleep(0.1)
+def clean_default_db(*, modified_after=None):
+    """Clean (vacuum) the default db if it is SQLite
+
+    When *modified_after* is set, database is cleaned only when it was modified
+    since the *modified_after* time.
+    """
+    # Limiting usage of in other function by lazy-imports.
+    # pylint: disable=import-outside-toplevel
+    import grass.script as gs
+
+    conn = gs.db_connection()
+    if conn and conn["driver"] != "sqlite":
+        return
+    # check if db exists
+    gis_env = gs.gisenv()
+    database = conn["database"]
+    database = database.replace("$GISDBASE", gis_env["GISDBASE"])
+    database = database.replace("$LOCATION_NAME", gis_env["LOCATION_NAME"])
+    database = database.replace("$MAPSET", gis_env["MAPSET"])
+    database = Path(database)
+    if not database.is_file():
+        return
+    file_stat = database.stat()
+    # Small size based on MEMORYMB (MiB) or its default.
+    small_db_size = int(gis_env.get("MEMORYMB", 300)) * (1 << 20)
+    if file_stat.st_size <= small_db_size:
+        return
+    if modified_after:
+        modified_time = datetime.datetime.fromtimestamp(
+            file_stat.st_mtime, tz=datetime.timezone.utc
+        )
+        if modified_after >= modified_time:
+            return
+    process = gs.start_command("db.execute", sql="VACUUM")
+    gs.verbose(_("Cleaning up default SQLite database..."))
+    process.wait()
```

Comment:
Almost the same. The message job runs in parallel. Because the message is a subprocess, it takes some tiny but measurable time to execute, so I tried to shave off some time by showing it in parallel while the vacuum is running. I tried to clarify that now in the comment.

With the greater threshold for vacuum I introduced later, the practical difference is likely negligible, but doesn't do any harm. It breaks the promise of no multiple processes, but not in a data processing context, so likely permissible. Anyway, I can remove it if it seems too strange.